File: tests/python/test_decoration.py
"""DecoratedInterval: decoration propagation and the polymorphic math API."""

import pickle

import pytest

import pyintval as iv

DI = iv.DecoratedInterval


def test_construction_decorations():
    assert DI(1, 2).decoration == "com"
    assert DI(0, float("inf")).decoration == "dac"
    assert DI("[entire]").decoration == "dac"
    assert DI("[empty]").decoration == "trv"
    assert DI.nai().decoration == "ill"
    assert DI.nai().is_nai


def test_from_parts_and_properties():
    d = DI.from_parts(iv.Interval(1, 2), "def")
    assert d.decoration == "def"
    assert d.interval == iv.Interval(1, 2)
    assert d.lo == 1.0 and d.hi == 2.0
    assert d.is_defined and not d.is_defined_and_continuous and not d.is_common
    with pytest.raises(ValueError):
        DI.from_parts(iv.Interval(1, 2), "bogus")


def test_predicate_ladder():
    com = DI(1, 2)
    assert com.is_common and com.is_defined_and_continuous and com.is_defined
    dac = DI(0, float("inf"))
    assert not dac.is_common and dac.is_defined_and_continuous and dac.is_defined
    trv = iv.log(DI(0, 2))
    assert not trv.is_defined and not trv.is_defined_and_continuous


def test_continuous_ops_keep_com():
    a, b = DI(1, 2), DI(3, 4)
    assert (a + b).decoration == "com"
    assert (a * b).decoration == "com"
    assert iv.exp(a).decoration == "com"
    assert iv.sin(a).decoration == "com"
    assert iv.sqrt(DI(1, 4)).decoration == "com"


def test_domain_violations_give_trv():
    assert iv.sqrt(DI(-1, 4)).decoration == "trv"
    assert iv.log(DI(0, 2)).decoration == "trv"
    assert (DI(1, 2) / DI(-1, 1)).decoration == "trv"
    assert iv.asin(DI(-2, 0.5)).decoration == "trv"
    assert iv.tan(DI(1.5, 1.7)).decoration == "trv"
    assert iv.atan2(DI(-1, 1), DI(-1, 1)).decoration == "trv"  # origin


def test_step_functions_and_branch_cut():
    assert iv.floor(DI(0.2, 0.8)).decoration == "com"
    assert iv.floor(DI(0.5, 1.5)).decoration == "def"
    assert iv.sign(DI(1, 2)).decoration == "com"
    assert iv.sign(DI(-1, 1)).decoration == "def"
    assert iv.atan2(DI(-1, 1), DI(-2, -1)).decoration == "def"  # branch cut


def test_composition_propagates_weakest():
    # A single domain violation poisons the whole certificate.
    assert iv.exp(iv.sqrt(DI(-1, 4))).decoration == "trv"
    assert (iv.log(DI(0, 2)) + DI(1, 2)).decoration == "trv"
    # A clean composition certifies defined & continuous.
    f = iv.sin(DI(0, 1)) / (iv.sqr(DI(0, 1)) + DI(1, 1))
    assert f.is_defined_and_continuous
    # ill poisons regardless.
    assert (DI.nai() + DI(1, 2)).decoration == "ill"


def test_operators_and_promotion():
    a = DI(2, 3)
    assert (a + 1).decoration == "com"
    assert (2 * a).interval == iv.Interval(4, 6)
    assert (a + iv.Interval(1, 1)).decoration == "com"  # bare interval promoted
    assert (a**2).interval == iv.Interval(4, 9)
    assert 2**0.5 in (DI(2) ** 0.5).interval
    assert (-a).interval == iv.Interval(-3, -2)


def test_polymorphism_preserves_type():
    assert isinstance(iv.exp(iv.Interval(0, 1)), iv.Interval)
    assert isinstance(iv.exp(DI(0, 1)), DI)
    assert isinstance(iv.atan2(DI(1, 2), DI(1, 2)), DI)
    assert isinstance(iv.hypot(iv.Interval(3), iv.Interval(4)), iv.Interval)


def test_repr_str_pickle():
    a = DI(1, 2)
    assert "com" in repr(a)
    assert str(a).endswith("_com")
    assert pickle.loads(pickle.dumps(a)) == a
    assert eval(repr(a), {"DecoratedInterval": DI, "Interval": iv.Interval}) == a


def test_decorated_set_cancel_reverse_ops():
    # F3: set operations, cancellative subtraction, and reverse ops accept
    # decorated intervals; the result is the bare op decorated trv, and NaI
    # propagates.
    I = iv.Interval
    fp = DI.from_parts
    a, b = fp(I(1, 3), "com"), fp(I(2.1, 4), "com")
    assert iv.intersection(a, b) == fp(I(2.1, 3), "trv")
    assert iv.hull(a, b) == fp(I(1, 4), "trv")
    assert iv.cancel_plus(fp(I(-1, 3), "com"), fp(I(0, 1), "com")).decoration == "trv"
    assert iv.sqr_rev(fp(I(0, 25), "def")) == fp(I(-5, 5), "trv")
    assert iv.mul_rev(a, b).decoration == "trv"
    assert iv.abs_rev(fp(I(-1, 2), "com")).decoration == "trv"
    # NaI poisons the result.
    assert iv.intersection(DI.nai(), a).is_nai
    assert iv.mul_rev(DI.nai(), b).is_nai
    # Bare Interval overloads still work unchanged.
    assert iv.intersection(I(1, 3), I(2, 4)) == I(2, 3)


def test_decorated_text_literals():
    # F4: native "[a,b]_dec" parsing; no suffix -> the tightest decoration; the
    # uncertain form works decorated too.
    assert DI("[1,1e3]_com") == DI.from_parts(iv.Interval(1, 1000), "com")
    assert DI("[1,1E3]_COM").decoration == "com"  # case-insensitive suffix
    assert DI("[entire]").decoration == "dac"
    assert DI("3.56?1e2") == DI.from_parts(iv.Interval(355, 357), "com")
    # Invalid decoration -> NaI (over-claim, unknown, or 'ill'); "[nai]" -> NaI.
    assert DI("[nai]").is_nai
    assert DI("[empty]_com").is_nai
    assert DI("[,]_com").is_nai
    assert DI("[1,2]_ill").is_nai
    assert DI("[1,2]_fooo").is_nai
    # Genuinely malformed still raises.
    with pytest.raises(ValueError):
        DI("garbage")

File: tests/cpp/test_decoration.cpp
// Tests for the IEEE 1788 decoration system: that each operation reports the
// correct local decoration, and that composition propagates the weakest one.

#include "doctest/doctest.h"
#include "pyintval/decoration.hpp"

using namespace pyintval;

namespace {
Decoration dec_of(const DecoratedInterval& d) { return d.dec; }
DecoratedInterval D(double lo, double hi) { return decorate(make(lo, hi)); }
}  // namespace

TEST_CASE("construction assigns the tightest decoration") {
  CHECK(D(1.0, 2.0).dec == Decoration::com);                       // bounded, nonempty
  CHECK(decorate(make(0.0, detail::kInf)).dec == Decoration::dac);  // unbounded
  CHECK(decorate(entire()).dec == Decoration::dac);
  CHECK(decorate(empty()).dec == Decoration::trv);
  CHECK(nai().dec == Decoration::ill);
}

TEST_CASE("continuous operations preserve com/dac") {
  CHECK(add(D(1, 2), D(3, 4)).dec == Decoration::com);
  CHECK(mul(D(1, 2), D(3, 4)).dec == Decoration::com);
  CHECK(exp(D(0, 1)).dec == Decoration::com);
  CHECK(sin(D(0, 1)).dec == Decoration::com);
  CHECK(sqr(D(-2, 3)).dec == Decoration::com);
  // An unbounded (but still defined & continuous) input drops com to dac.
  CHECK(add(D(1, 2), decorate(make(0.0, detail::kInf))).dec == Decoration::dac);
  CHECK(exp(decorate(make(-detail::kInf, 0.0))).dec == Decoration::dac);
}

TEST_CASE("square root domain") {
  CHECK(sqrt(D(1, 4)).dec == Decoration::com);   // strictly inside domain
  CHECK(sqrt(D(0, 4)).dec == Decoration::com);   // 0 is in the domain, continuous
  CHECK(sqrt(D(-1, 4)).dec == Decoration::trv);  // domain violated
  CHECK(sqrt(D(-4, -1)).dec == Decoration::trv);  // fully outside -> empty, trv
}

TEST_CASE("logarithms require strictly positive input") {
  CHECK(log(D(1, 2)).dec == Decoration::com);
  CHECK(log(D(0, 2)).dec == Decoration::trv);   // log undefined at 0
  CHECK(log(D(-1, 2)).dec == Decoration::trv);
  CHECK(log1p(D(-0.5, 2.0)).dec == Decoration::com);
  CHECK(log1p(D(-1.0, 2.0)).dec == Decoration::trv);
}

TEST_CASE("division and reciprocal by a zero-containing divisor are trv") {
  CHECK(div(D(1, 2), D(3, 4)).dec == Decoration::com);
  CHECK(div(D(1, 2), D(-1, 1)).dec == Decoration::trv);
  CHECK(div(D(1, 2), D(0, 1)).dec == Decoration::trv);
  CHECK(recip(D(2, 4)).dec == Decoration::com);
  CHECK(recip(D(-1, 1)).dec == Decoration::trv);
}

TEST_CASE("inverse trig and hyperbolic domains") {
  CHECK(asin(D(-0.5, 0.5)).dec == Decoration::com);
  CHECK(asin(D(-1.0, 1.0)).dec == Decoration::com);   // closed domain endpoints ok
  CHECK(asin(D(-2.0, 0.5)).dec == Decoration::trv);
  CHECK(acos(D(0.0, 1.0)).dec == Decoration::com);
  CHECK(acos(D(-1.5, 0.0)).dec == Decoration::trv);
  CHECK(acosh(D(1.0, 3.0)).dec == Decoration::com);
  CHECK(acosh(D(0.5, 3.0)).dec == Decoration::trv);
  CHECK(atanh(D(-0.5, 0.5)).dec == Decoration::com);
  CHECK(atanh(D(-1.0, 0.5)).dec == Decoration::trv);   // open domain endpoint
}

TEST_CASE("tan asymptotes drop to trv") {
  CHECK(tan(D(0.1, 0.2)).dec == Decoration::com);
  CHECK(tan(D(1.5, 1.7)).dec == Decoration::trv);   // straddles pi/2
  CHECK(tan(D(0.0, 4.0)).dec == Decoration::trv);   // spans a full period
}

TEST_CASE("step functions are def unless locally constant") {
  CHECK(floor(D(0.2, 0.8)).dec == Decoration::com);   // constant on the cell
  CHECK(floor(D(0.5, 1.5)).dec == Decoration::def);   // crosses an integer
  CHECK(ceil(D(0.2, 0.8)).dec == Decoration::com);
  CHECK(trunc(D(-0.5, 0.5)).dec == Decoration::com);  // trunc == 0 on (-1,1): continuous
  CHECK(trunc(D(0.5, 1.5)).dec == Decoration::def);   // crosses the jump at 1
  CHECK(sign(D(1.0, 2.0)).dec == Decoration::com);
  CHECK(sign(D(-1.0, 1.0)).dec == Decoration::def);
}

TEST_CASE("atan2 origin, branch cut, and clean regions") {
  CHECK(atan2(D(1, 2), D(1, 2)).dec == Decoration::com);       // first quadrant
  CHECK(atan2(D(-1, 1), D(-2, -1)).dec == Decoration::def);    // crosses branch cut
  CHECK(atan2(D(-1, 1), D(-1, 1)).dec == Decoration::trv);     // contains origin
  CHECK(hypot(D(3, 4), D(-1, 2)).dec == Decoration::com);      // continuous everywhere
}

TEST_CASE("pow domain") {
  CHECK(pow(D(2, 3), decorate(point(0.5))).dec == Decoration::com);   // positive base
  CHECK(pow(D(-1, 3), decorate(point(0.5))).dec == Decoration::trv);  // base can be < 0
  CHECK(pow(D(-2, 3), decorate(point(2.0))).dec == Decoration::com);  // integer exponent
  CHECK(pow(D(-2, 3), decorate(point(-1.0))).dec == Decoration::trv);  // n<0 with 0 in base
}

TEST_CASE("composition propagates the weakest decoration") {
  // A single domain violation anywhere poisons the whole certificate.
  CHECK(exp(sqrt(D(-1, 4))).dec == Decoration::trv);
  CHECK(add(log(D(0, 2)), D(1, 2)).dec == Decoration::trv);
  // A clean composition certifies defined-and-continuous.
  DecoratedInterval f = div(sin(D(0, 1)), add(sqr(D(0, 1)), decorate(point(1.0))));
  CHECK(static_cast<int>(f.dec) >= static_cast<int>(Decoration::dac));
  // ill poisons regardless of the other operand.
  CHECK(add(nai(), D(1, 2)).dec == Decoration::ill);
  CHECK(mul(D(1, 2), nai()).dec == Decoration::ill);
}
